--- a/drivers/net/bonding/bond_main.c
+++ b/drivers/net/bonding/bond_main.c
@@ -93,7 +93,10 @@
 #include <net/xdp.h>
 
 #include "bonding_priv.h"
-
+#if IS_ENABLED(CONFIG_NF_FLOW_TABLE)
+#include <linux/netfilter.h>
+#include <net/netfilter/nf_flow_table.h>
+#endif
 /*---------------------------- Module parameters ----------------------------*/
 
 /* monitor all links that often (in milliseconds). <=0 disables monitoring */
@@ -5883,6 +5886,35 @@ out:
 	return ret;
 }
 
+#if IS_ENABLED(CONFIG_NF_FLOW_TABLE)
+static int bond_dev_flow_offload_check(struct flow_offload_hw_path *path)
+{
+       struct net_device *master_dev = path->dev;
+       struct net_device *slave_dev[10];
+       struct list_head *iter;
+       int i = 0;
+       netdev_for_each_lower_dev(master_dev, slave_dev[i], iter) {
+               /* Check the link status of the slave device */
+       if (!(slave_dev[i]->flags & IFF_UP) ||
+               !netif_carrier_ok(slave_dev[i]))
+               continue;
+                       i++;
+       if (i >= ARRAY_SIZE(slave_dev))
+                       break;
+       }
+       if (i > 0) {
+               i = ((path->skb_hash) >> 1) % i;
+                       if (i >= 0 && i < ARRAY_SIZE(slave_dev)) {
+                       /* Choose a subordinate device by hash index */
+                       path->dev = slave_dev[i];
+                       }
+       }
+       if (path->dev->netdev_ops->ndo_flow_offload_check)
+               return path->dev->netdev_ops->ndo_flow_offload_check(path);
+       return 0;
+}
+#endif /* CONFIG_NF_FLOW_TABLE */
+
 static const struct ethtool_ops bond_ethtool_ops = {
 	.get_drvinfo		= bond_ethtool_get_drvinfo,
 	.get_link		= ethtool_op_get_link,
@@ -5924,6 +5956,9 @@ static const struct net_device_ops bond_
 	.ndo_xdp_get_xmit_slave = bond_xdp_get_xmit_slave,
 	.ndo_hwtstamp_get	= bond_hwtstamp_get,
 	.ndo_hwtstamp_set	= bond_hwtstamp_set,
+#if IS_ENABLED(CONFIG_NF_FLOW_TABLE)
+       .ndo_flow_offload_check = bond_dev_flow_offload_check,
+#endif
 };
 
 static const struct device_type bond_type = {
--- a/net/netfilter/nft_counter.c
+++ b/net/netfilter/nft_counter.c
@@ -15,6 +15,28 @@
 #include <net/netfilter/nf_tables.h>
 #include <net/netfilter/nf_tables_core.h>
 #include <net/netfilter/nf_tables_offload.h>
+#include <net/netfilter/nf_conntrack.h>
+#include <net/netfilter/nf_conntrack_acct.h>
+static struct hnat_desc {
+       u32 entry : 14;
+       #if defined(CONFIG_MEDIATEK_NETSYS_RX_V2)
+       u32 fill : 4;
+       #endif
+       u32 crsn : 5;
+       u32 sport : 4;
+       u32 alg : 1;
+       u32 iface : 8;
+       u32 ppe : 1;
+       u32 filled : 3;
+       u32 magic_tag_protect : 16;
+       u32 wdmaid : 8;
+       u32 rxid : 2;
+       u32 wcid : 10;
+       u32 bssid : 6;
+ } __packed;
+#define HIT_BIND_KEEPALIVE_DUP_OLD_HDR 0x15
+#define skb_hnat_reason(skb) (((struct hnat_desc *)(skb->head))->crsn)
+
 
 struct nft_counter {
 	u64_stats_t	bytes;
@@ -38,6 +60,24 @@ static inline void nft_counter_do_eval(s
 {
 	struct u64_stats_sync *nft_sync;
 	struct nft_counter *this_cpu;
+	struct nf_conn *ct;
+        struct nf_conn_acct *acct;
+        struct nf_conn_counter *conn_counter;
+        enum ip_conntrack_info ctinfo;
+	s64 bytes = 0;
+        s64 packets = 0;
+
+	if (unlikely(skb_hnat_reason(pkt->skb) == HIT_BIND_KEEPALIVE_DUP_OLD_HDR)){
+               ct = nf_ct_get(pkt->skb, &ctinfo);
+               if (ct) {
+                       acct = nf_conn_acct_find(ct);
+                       if (acct) {
+                       conn_counter = acct->counter;
+                       bytes = atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_bytes);
+                       packets = atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_packets);
+                       }
+               }
+       }
 
 	local_bh_disable();
 	this_cpu = this_cpu_ptr(priv->counter);
@@ -46,6 +86,8 @@ static inline void nft_counter_do_eval(s
 	u64_stats_update_begin(nft_sync);
 	u64_stats_add(&this_cpu->bytes, pkt->skb->len);
 	u64_stats_inc(&this_cpu->packets);
+	u64_stats_add(&this_cpu->bytes, bytes);
+	u64_stats_add(&this_cpu->packets, packets);
 	u64_stats_update_end(nft_sync);
 
 	local_bh_enable();
--- a/include/net/netfilter/nf_flow_table.h
+++ b/include/net/netfilter/nf_flow_table.h
@@ -196,6 +196,7 @@ struct flow_offload_hw_path {
 	u16 vlan_id;
 	u16 pppoe_sid;
 	u16 dsa_port;
+	u32 skb_hash;
 };
 
 #define NF_FLOW_TIMEOUT (30 * HZ)
--- a/include/net/netfilter/nf_conntrack_acct.h
+++ b/include/net/netfilter/nf_conntrack_acct.h
@@ -14,6 +14,8 @@
 struct nf_conn_counter {
 	atomic64_t packets;
 	atomic64_t bytes;
+	atomic64_t diff_packets;
+	atomic64_t diff_bytes;
 };
 
 struct nf_conn_acct {
--- a/net/ipv4/netfilter/ip_tables.c
+++ b/net/ipv4/netfilter/ip_tables.c
@@ -26,6 +26,8 @@
 #include <linux/netfilter_ipv4/ip_tables.h>
 #include <net/netfilter/nf_log.h>
 #include "../../netfilter/xt_repldata.h"
+#include <net/netfilter/nf_conntrack.h>
+#include <net/netfilter/nf_conntrack_acct.h>
 
 MODULE_LICENSE("GPL");
 MODULE_AUTHOR("Netfilter Core Team <coreteam@netfilter.org>");
@@ -271,6 +273,25 @@ ipt_handle_default_rule(struct ipt_entry
 	return true;
 }
 
+static struct hnat_desc {
+	u32 entry : 14;
+	#if defined(CONFIG_MEDIATEK_NETSYS_RX_V2)
+	u32 fill : 4;
+	#endif
+	u32 crsn : 5;
+	u32 sport : 4;
+	u32 alg : 1;
+	u32 iface : 8;
+	u32 ppe : 1;
+	u32 filled : 3;
+	u32 magic_tag_protect : 16;
+	u32 wdmaid : 8;
+	u32 rxid : 2;
+	u32 wcid : 10;
+	u32 bssid : 6;
+} __packed;
+#define HIT_BIND_KEEPALIVE_DUP_OLD_HDR 0x15
+#define skb_hnat_reason(skb) (((struct hnat_desc *)(skb->head))->crsn)
 /* Returns one of the generic firewall policies, like NF_ACCEPT. */
 unsigned int
 ipt_do_table(void *priv,
@@ -290,7 +311,12 @@ ipt_do_table(void *priv,
 	const struct xt_table_info *private;
 	struct xt_action_param acpar;
 	unsigned int addend;
-
+	
+	struct nf_conn *ct;
+	struct nf_conn_acct *acct;
+	struct nf_conn_counter *conn_counter;
+	enum ip_conntrack_info ctinfo;
+	
 	/* Initialization */
 	WARN_ON(!(table->valid_hooks & (1 << hook)));
 	local_bh_disable();
@@ -301,9 +327,22 @@ ipt_do_table(void *priv,
 	e = get_entry(table_base, private->hook_entry[hook]);
 	if (ipt_handle_default_rule(e, &verdict)) {
 		struct xt_counters *counter;
-
 		counter = xt_get_this_cpu_counter(&e->counters);
 		ADD_COUNTER(*counter, skb->len, 1);
+		
+		if (unlikely(skb_hnat_reason(skb) == HIT_BIND_KEEPALIVE_DUP_OLD_HDR)){
+		ct = nf_ct_get(skb, &ctinfo);
+		if (ct) {
+			acct = nf_conn_acct_find(ct);
+			if (acct) {
+			conn_counter = acct->counter;			
+			ADD_COUNTER(*counter, 		   	    
+ atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_bytes),atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_packets));
+				  }
+			}
+		}
+
+			
 		local_bh_enable();
 		return verdict;
 	}
@@ -359,6 +398,17 @@ ipt_do_table(void *priv,
 
 		counter = xt_get_this_cpu_counter(&e->counters);
 		ADD_COUNTER(*counter, skb->len, 1);
+		if (unlikely(skb_hnat_reason(skb) == HIT_BIND_KEEPALIVE_DUP_OLD_HDR)){
+		ct = nf_ct_get(skb, &ctinfo);
+		if (ct) {
+			acct = nf_conn_acct_find(ct);
+			if (acct) {
+			conn_counter = acct->counter;			
+			ADD_COUNTER(*counter, 		   	    
+ atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_bytes),atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_packets));
+				  }
+			}
+		}
 
 		t = ipt_get_target_c(e);
 		WARN_ON(!t->u.kernel.target);
--- a/net/ipv6/netfilter/ip6_tables.c
+++ b/net/ipv6/netfilter/ip6_tables.c
@@ -30,6 +30,8 @@
 #include <linux/netfilter/x_tables.h>
 #include <net/netfilter/nf_log.h>
 #include "../../netfilter/xt_repldata.h"
+#include <net/netfilter/nf_conntrack.h>
+#include <net/netfilter/nf_conntrack_acct.h>
 
 MODULE_LICENSE("GPL");
 MODULE_AUTHOR("Netfilter Core Team <coreteam@netfilter.org>");
@@ -243,6 +245,25 @@ ip6t_next_entry(const struct ip6t_entry
 	return (void *)entry + entry->next_offset;
 }
 
+static struct hnat_desc {
+	u32 entry : 14;
+	#if defined(CONFIG_MEDIATEK_NETSYS_RX_V2)
+	u32 fill : 4;
+	#endif
+	u32 crsn : 5;
+	u32 sport : 4;
+	u32 alg : 1;
+	u32 iface : 8;
+	u32 ppe : 1;
+	u32 filled : 3;
+	u32 magic_tag_protect : 16;
+	u32 wdmaid : 8;
+	u32 rxid : 2;
+	u32 wcid : 10;
+	u32 bssid : 6;
+} __packed;
+#define HIT_BIND_KEEPALIVE_DUP_OLD_HDR 0x15
+#define skb_hnat_reason(skb) (((struct hnat_desc *)(skb->head))->crsn)
 /* Returns one of the generic firewall policies, like NF_ACCEPT. */
 unsigned int
 ip6t_do_table(void *priv, struct sk_buff *skb,
@@ -260,7 +281,12 @@ ip6t_do_table(void *priv, struct sk_buff
 	const struct xt_table_info *private;
 	struct xt_action_param acpar;
 	unsigned int addend;
-
+	
+	struct nf_conn *ct;
+	struct nf_conn_acct *acct;
+	struct nf_conn_counter *conn_counter;
+	enum ip_conntrack_info ctinfo;
+	
 	/* Initialization */
 	stackidx = 0;
 	indev = state->in ? state->in->name : nulldevname;
@@ -283,6 +309,7 @@ ip6t_do_table(void *priv, struct sk_buff
 	cpu        = smp_processor_id();
 	table_base = private->entries;
 	jumpstack  = (struct ip6t_entry **)private->jumpstack[cpu];
+	
 
 	/* Switch to alternate jumpstack if we're being invoked via TEE.
 	 * TEE issues XT_CONTINUE verdict on original skb so we must not
@@ -319,7 +346,20 @@ ip6t_do_table(void *priv, struct sk_buff
 
 		counter = xt_get_this_cpu_counter(&e->counters);
 		ADD_COUNTER(*counter, skb->len, 1);
+		
+		if (unlikely(skb_hnat_reason(skb) == HIT_BIND_KEEPALIVE_DUP_OLD_HDR)){
+		ct = nf_ct_get(skb, &ctinfo);
+		if (ct) {
+			acct = nf_conn_acct_find(ct);
+			if (acct) {
+			conn_counter = acct->counter;			
+			ADD_COUNTER(*counter, 		   	    
+ atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_bytes),atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_packets));
+				  }
+			}
+		}
 
+			
 		t = ip6t_get_target_c(e);
 		WARN_ON(!t->u.kernel.target);
 
